### DISCUSSION TITLE
Let's discuss pedagogy of the notebook tutorial

### DISCUSSION THREAD

**Discussion by beckynevin** [lsst-epo/citizen-science-notebooks#42] [General]
@bnord and @ericdrosas87 
Question - Currently the manifest part of the Citizen Science Testing notebook is a little hard to follow for a new user. I'm particularly wondering about how we want to approach the 'fields_to_add' which is mentioned here but does not appear in the code
<img width="756" alt="image" src="https://github.com/lsst-epo/citizen-science-notebooks/assets/26393168/47bf50d9-9ee8-420c-b0ea-7d4958221279">
I'm working on all of this in branch clarifying_pedagogy

**Comment by beckynevin** [comment]
@bnord 
Another question about the manifest file sending - we mention RFC4180, which seems like technical jargon (I'm not even sure what it stands for). Is there a way to hide this sort of technical detail a little better within jupyter notebooks? Like some sort of feature you can click on which then drops down an additional block quote?
<img width="735" alt="image" src="https://github.com/lsst-epo/citizen-science-notebooks/assets/26393168/dede654b-1b23-4369-915d-b7e865da30c3">

**Reply by bnord** [reply]
- I think we should remove everything about what the user doesn't need to care about. we can put it in some appendix in case they get curious. 
- The guidelines required by RFC4180 should be made user-friendly. I don't think people should need to read a detailed document. but, I don't know what's in RFC4180. Do you have a link? I couldn't find one online.
-

**Reply by ericdrosas87** [reply]
[RFC](https://en.wikipedia.org/wiki/Request_for_Comments) = Request For Comments and is an acronym used when a document is created to formalize something, in this case: formalizing what a CSV (comma separated value) file format is. I just included added it to the notebook for reference because the `metadata.csv` must be in CSV format. Feel free to remove if it doesn't make sense to have it here.

Edit: Changed link to Wikipedia article

**Reply by clareh** [reply]
I am in favor of removing this chunk - I think CSV formats are common enough and if we really want, we can add this into a "trouble shooting" part of the PI guide

**Comment by bnord** [comment]
> @bnord and @ericdrosas87 Question - Currently the manifest part of the Citizen Science Testing notebook is a little hard to follow for a new user. I'm particularly wondering about how we want to approach the 'fields_to_add' which is mentioned here but does not appear in the code <img alt="image" width="756" src="https://user-images.githubusercontent.com/26393168/239617901-47bf50d9-9ee8-420c-b0ea-7d4958221279.png"> I'm working on all of this in branch clarifying_pedagogy

Where would the fields to add actually get used?

**Reply by ericdrosas87** [reply]
The Zooniverse platform will parse out the column names and if it identifies some Zooniverse-specific syntax for a column it may do something special with it (hide it from being visible to the citizen scientists, etc.). For image data, the fields are mostly important because they can be used to associate the output/classified data that comes from a Zooniverse completed subject set with the original data. For tabular data, you can expect these fields to be shown to the citizen scientists on the Zooniverse website.

**Reply by beckynevin** [reply]
We ended up creating a function called 'make_manifest_and_images' within utils.py that packages up this chunk of code. Inputs are: results_table, butler, batch_dir - this function will create and save the data (results_table) into a manifest file to the batch_dir.

**Comment by beckynevin** [comment]
@ericdrosas87 and @clareh, Brian, Sree and I met today and we have a proposal for how to better organize the tutorial notebooks. What do you think about condensing some of the cells in Citizen_Science_Testing.ipynb by creating separate python scripts, i.e. create_manifest.py that accept various arguments but have a default set of arguments. Also, we're wondering what to do with Citizen_Science_SDK.ipynb - should it be a script as well? Discuss.

**Reply by bnord** [reply]
I think we should script and package as much as possible, lest we invite questions about things that users shouldn't even need to worry about.

**Reply by bnord** [reply]
Is it possible to get into package release schedule of other Rubin software products?

**Reply by ericdrosas87** [reply]
@bnord I was actually planning on eventually consolidating the various scripts and SDK notebook and publishing it as a package to PyPI so that it can be installed via `pip`. That way aren't necessarily tied to DM's release schedule for the RSP Notebook Aspect. We should probably assess whether now or when is a good time to do that - ideally it should happen when most of the iterative development reaches a stable point. Will there be frequent/many changes soon that we should wait for? Should we just go for it and publish the package to iron out the installation details now? Those are the sort of questions I have in mind before doing so.

**Reply by bnord** [reply]
I think we should start planning that now, because if we leave it as it is, we'll need to put effort into explaining them in a notebook, and I think that would be an inefficient use of effort.

**Reply by bnord** [reply]
Even if things change later, I think we can iterate in the script itself. We can do different software releases. Software always changes.

**Reply by ericdrosas87** [reply]
>because if we leave it as it is, we'll need to put effort into explaining them in a notebook

Makes sense, I'll start looking into the PyPI publish/packaging process

**Reply by bnord** [reply]
@ericdrosas87 our group has done this a little bit with poetry, but we're noobs to it. Poetry is what we're most familiar with for creating a package. This is to say, we're not yet conversant here, but we're learning.

**Reply by beckynevin** [reply]
I can chime in and say poetry is great - it even has a publish command: [poetry publish](https://python-poetry.org/docs/libraries/#:~:text=Poetry%20will%20publish%20to%20PyPI,publish%20on%20PyPI%20as%20well.)

**Reply by bnord** [reply]
How would y'all like to organize this? Should one or two people design modules/scripts and then get feedback? Should we all design together?

**Reply by ericdrosas87** [reply]
I've never dabbled in Poetry usage, but I'm not opposed. I have used Setuptools in the past because it's bundled with many Python distributions, but I'm open to whatever works best for everyone. Although this does bring an interesting question to mind: Who will be maintaining the package after it's initially developed? EPO devs (for now just me)? Brian/Becky and co? All of us + Clare? Once that question is settled, I think then it becomes a matter of what everyone is most comfortable with and what makes the most sense (software), and then we can design from there.

**Reply by clareh** [reply]
Just catching up on your discussion here - totally agree with the script/package approach. However, the only caveat we need to keep in mind is to not sacrifice flexibility for simplicity (I don't think anything proposed here crosses that line, but just to keep in mind!!)

**Reply by bnord** [reply]
1. we also may not need to sacrifice much flexibility when scripting. 
2. the code will also be open-source, so people can always pull out what they need and make their own thing.

**Comment by beckynevin** [comment]
Okay I have an updated version of Citizen_Science_Testing.ipynb that I'd like feedback on - https://github.com/lsst-epo/citizen-science-notebooks/blob/clarifying_pedagogy/Citizen_Science_Testing.ipynb
@bnord @ericdrosas87 @clareh @jsv1206

**Reply by bnord** [reply]
Is the variable naming convention consistent? e.g., `slugName` 
e.g., https://www.theserverside.com/feature/A-guide-to-common-variable-naming-conventions#:~:text=The%20standard%20naming%20conventions%20used,snake%20case

**Reply by bnord** [reply]
I think the approval committee is no longer this: `EPO Data Rights panel.`

**Reply by bnord** [reply]
Should we put a table of contents or outline at the top?

**Reply by bnord** [reply]
` Make a subject set to send` -->  `Make a subject set ` for `Make a set of subjects`

Make a subject set to send it's tough grammar to parse.

**Reply by bnord** [reply]
Can the manifest file be called `manifest.csv` instead `metadata.csv`?

or can we call it a "metadata file?" 

I'm wondering why the name of the file must be hardcoded as a particular word, but then we refer to that file with not that same word. 

I imagine this could invite minor confusion.

**Reply by bnord** [reply]
This line is probably unnecessary
`You are welcome to edit the automatically created manifest file (option 1) or create a new manifest file (option 2).`

**Reply by bnord** [reply]
`Try not to reuse names.`

why? we should probably explain the consequences.

**Reply by bnord** [reply]
What does `one of each export `mean?

Is that export per project, one export per day, one export related manifest file?

**Reply by clareh** [reply]
I think we should stick to the word manifest as that is what zooniverse uses to talk about that file in their documentation (consistency!).

**Reply by beckynevin** [reply]
there are a couple of different choices for export data type on the zooniverse site - for instance, you could export details about the workflow or you could export the actual user classifications. Maybe we don't need to specify this?

**Comment by clareh** [comment]
Some thoughts:
- I think it would be good to include a "You can go look at the cutouts you made in this folder" explanation right before Part 3.0

- Can this be followed by a "If you want to change where data is saved, go here"

- "Query can be modified to select other types of sources." This needs some direction on how to do this and I think we should point again to the tutorial notebook that exists which explains butler queries. (This is not an easy thing to change if you are new to all this!)

- "The below cell writes the metadata.csv file to the filesystem, which will be used by Zooniverse." we should link to the Zooniverse documentation about manifest files (and definitely call them manifest files!! otherwise its going to get confusing). Emphasize that this file can be used to initiate options on the Zooniverse side (which we won't get into here but I think we need to acknowledge they exist)

**Reply by beckynevin** [reply]
I addressed all of these comments in the version that is now merged into main.

**Comment by beckynevin** [comment]
Okay new question that is on my mind that is related to our connection with Zooniverse.

@clareh @ericdrosas87 @jsv1206 @bnord 

This came up as we've begun putting together the PCW tutorial plan. When users get to the step of retrieving classifications, do they have to be added to a project to retrieve classifications from Zooniverse? Do we want them to do this with their own project? I see several options for what to do at this point in the tutorial, let me know what you think:
1. Skip this step for other people and just show an example on screen of us running the retrieval cell for the project we already have access to.
2. Is there some way to add all the people to our project or have it be fully public so that everyone can use the same project ID to download classifications?
3. Have everyone create individual projects and take time during the tutorial to classify one another's galaxies (this seems a little beside the point of the tutorial). Then, download classifications from your own project. This also seems problematic because I'm not sure you're immediately able to download classifications.

**Reply by bnord** [reply]
> When users get to the step of retrieving classifications, do they have to be added to a project to retrieve classifications from Zooniverse? Do we want them to do this with their own project?

Is this a non-automatic or time-expensive process?

**Reply by bnord** [reply]
> Skip this step for other people

who's other people? who's non-other people?

**Reply by ericdrosas87** [reply]
> do they have to be added to a project to retrieve classifications from Zooniverse?

If I'm remembering correctly, then yes they need to be added to the project on the Zooniverse platform as a "Collaborator", but we should get confirmation on that from Chris Lintott

**Reply by bnord** [reply]
how automated is the procedure of being added as a collaborator? Is ChrisL in this github org? can we tag him?

**Reply by ericdrosas87** [reply]
> Is ChrisL in this github org? can we tag him?

Oh no but @eatyourgreens is, we can probably ask him what role is required to programmatically fetch completed workflow data (classifications) or if another user can be programmatically added to another Zooniverse user's project

**Reply by eatyourgreens** [reply]
The web form that adds people to a project uses the Panoptes API, so the project owner should be able to programmatically add people. I’ll look up the exact API requests that are used. You probably need to know their Zooniverse login. 

I’m pretty sure the allowed roles to generate classification exports are project owner and collaborator. I imagine the exported CSV itself can be downloaded by anyone with the URL, while the URL is live.

**Reply by beckynevin** [reply]
@eatyourgreens do you mean that anyone can download the CSV via the Zooniverse website but that only the project owner and collaborators can download via our tutorial ipython notebook currently?

**Reply by eatyourgreens** [reply]
Sorry, I meant that only  the project owner and collaborators can request a new classifications CSV via the Panoptes API. In the project builder, we'll publish a link (and email them) once the CSV is ready. I don't think that link itself uses Panoptes authentication, as it's a link direct to the CSV file on panoptesuploads.blob.core.windows.net.

I haven't played around with the Python client for classifications exports, but I assume it must do something similar. Or at least, I assume the Python client request must be made by a member of the project team.

**Reply by beckynevin** [reply]
Ah okay so I think this answers another one of my questions - it's a manual process on the project builder side right? 

Also let me clarify my understanding so far on how retrieving classifications works, let me know if any of this is incorrect - 

- There are multiple ways to retrieve classifications from Zooniverse. 
- One way is to go physically click on the 'get classifications' button on the Zooniverse website. This may take a little while because it needs to be manually uploaded by the Zooniverse team. Only project owners and collaborators can do this.
- Using the panoptes API. This also publishes a csv in the project builder.
- It is also possible to get classifications via the python client (this this the way we do it in the tutorial notebook). Does this also require manual upload on the Zooniverse side? Can this be done if the project owner has not requested the csv on the Zooniverse or via the panoptes API first?

Also another question - when you say the project owners should be able to programatically add users to their project? Is there any sort of a delay associated with this such as authorization on the Zooniverse side? <-- we're wondering if we can quickly add users to a project so that they can quickly grab the classifications.

**Reply by eatyourgreens** [reply]
If you know a user’s login, I think the project owner (or a collaborator) can add them with an API call. Authentication is handled by passing your OAuth bearer token as the Authorization header with the API request. The Python client will take care of auth for you, but I don’t think it has classes yet for updating project teams. 

There’s no delay there. 

Classification exports are completely automated. The button in the project builder makes an API request to generate a new CSV. The delay is purely the time taken to gather the classifications and write the CSV to cloud storage. 

There are a couple of ways to speed up exports. One is to request classifications just for one workflow, or even for one subject set. Another way is to export incrementally, starting each export at a specific classification ID rather than requesting the whole set.

**Reply by eatyourgreens** [reply]
Actually, now that I write that, I think the delay in getting exports from the project builder also includes the time that your request sits in a queue, before it’s processed.

**Reply by eatyourgreens** [reply]
Would there be any value in publishing an example classifications CSV somewhere, then linking to that example from the tutorial. 

That would make the example classifications public, and avoid permissions problems.

**Reply by beckynevin** [reply]
I think this is a great idea. For now, we're just going to show ourselves accessing our own project in the PCW tutorial. So not an issue immediately, but could be a cool option if people want to see what it would look like to download classifications before they do their whole project on Zooniverse.
